File: 2025_fase2/solutions/estrada_alex.py
import sys
import heapq

INF = 10**18

def dijkstra(N, M, K, cypria_set, S, T, edges):
    # Grafo como lista de adjacência
    adj = [[] for _ in range(N + 1)]
    for u, v, w, p in edges:
        adj[u].append((v, w, p))
        adj[v].append((u, w, p))

    # dist[node][used_vigilancia][passou_cypria]
    dist = [[[INF] * 2 for _ in range(K + 1)] for _ in range(N + 1)]

    start_passed = 1 if S in cypria_set else 0
    dist[S][0][start_passed] = 0

    pq = [(0, S, 0, start_passed)]  # (dist, node, usadas, passou)
    
    while pq:
        d, u, used, passed = heapq.heappop(pq)
        if d != dist[u][used][passed]:
            continue

        for v, w, p in adj[u]:
            nused = used + p
            if nused > K:
                continue
            npassed = passed or (v in cypria_set)
            nd = d + w
            if nd < dist[v][nused][npassed]:
                dist[v][nused][npassed] = nd
                heapq.heappush(pq, (nd, v, nused, npassed))

    ans = min(dist[T][k][1] for k in range(K + 1))
    return -1 if ans >= INF else ans

def main():
    data = sys.stdin.read().strip().split()
    if not data:
        return
    it = iter(data)
    N = int(next(it))
    M = int(next(it))
    K = int(next(it))
    R = int(next(it))

    cypria_cities = [int(next(it)) for _ in range(R)]
    cypria_set = set(cypria_cities)

    S = int(next(it))
    T = int(next(it))

    edges = []
    for _ in range(M):
        u = int(next(it))
        v = int(next(it))
        w = int(next(it))
        p = int(next(it))
        edges.append((u, v, w, p))

    ans = dijkstra(N, M, K, cypria_set, S, T, edges)
    print(ans)

if __name__ == "__main__":
    main()

File: 2025_fase2/solutions/estrada_alex.java
import java.util.*;

public class estrada_alex {
    static class Edge {
        int to, w, p;
        Edge(int to, int w, int p) {
            this.to = to;
            this.w = w;
            this.p = p;
        }
    }

    static class State implements Comparable<State> {
        long dist;
        int node, used, passed;
        State(long dist, int node, int used, int passed) {
            this.dist = dist;
            this.node = node;
            this.used = used;
            this.passed = passed;
        }
        public int compareTo(State o) {
            return Long.compare(this.dist, o.dist);
        }
    }

    static final long INF = Long.MAX_VALUE;
    static final int MAXK = 35;
    static List<Edge>[] adj;
    static boolean[] inCypria;
    static long[][][] dist;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt();
        int M = sc.nextInt();
        int K = sc.nextInt();
        int R = sc.nextInt();

        inCypria = new boolean[N + 1];
        for (int i = 0; i < R; i++) {
            int c = sc.nextInt();
            inCypria[c] = true;
        }

        int S = sc.nextInt();
        int T = sc.nextInt();

        adj = new ArrayList[N + 1];
        for (int i = 1; i <= N; i++) adj[i] = new ArrayList<>();

        for (int i = 0; i < M; i++) {
            int u = sc.nextInt();
            int v = sc.nextInt();
            int w = sc.nextInt();
            int p = sc.nextInt();
            adj[u].add(new Edge(v, w, p));
            adj[v].add(new Edge(u, w, p));
        }

        dist = new long[N + 1][K + 1][2];
        for (int i = 1; i <= N; i++)
            for (int k = 0; k <= K; k++)
                Arrays.fill(dist[i][k], INF);

        int startPassed = inCypria[S] ? 1 : 0;
        dist[S][0][startPassed] = 0;

        PriorityQueue<State> pq = new PriorityQueue<>();
        pq.add(new State(0, S, 0, startPassed));

        while (!pq.isEmpty()) {
            State cur = pq.poll();
            if (cur.dist != dist[cur.node][cur.used][cur.passed]) continue;

            for (Edge e : adj[cur.node]) {
                int nused = cur.used + e.p;
                if (nused > K) continue;
                int npassed = cur.passed | (inCypria[e.to] ? 1 : 0);
                long nd = cur.dist + e.w;

                if (nd < dist[e.to][nused][npassed]) {
                    dist[e.to][nused][npassed] = nd;
                    pq.add(new State(nd, e.to, nused, npassed));
                }
            }
        }

        long ans = INF;
        for (int k = 0; k <= K; k++)
            ans = Math.min(ans, dist[T][k][1]); // 1 => passou por Cypria

        System.out.println(ans == INF ? -1 : ans);
        sc.close();
    }
}

